### ISSUE TITLE
400 error when creating a session without a password is not JSON

### ISSUE THREAD

**Issue by cainlevy** [FreedomBen/malan#95]
When I attempt to create a session but omit the password field, Malan responds with a 400 error that does not match my `Accept: application/json` header.

The response includes development debugging because I'm using `malan-arm-dev:latest`. I have not tried using a production build.

## What I Expected

I expected to either get JSON "bad request" errors or the standard "invalid credentials" error.

## Example

Notice that this curl request does not have a `"password"` field:

```
curl 'http://127.0.0.1:7102/api/sessions' -H 'Accept: application/json' -H 'Content-Type: application/json' --data-raw '{"session":{"username":"example"}}'
```

## Error

```
# Phoenix.ActionClauseError at POST /api/sessions

Exception:

    ** (Phoenix.ActionClauseError) no function clause matching in MalanWeb.SessionController.create/2
    
    The following arguments were given to MalanWeb.SessionController.create/2:
    
        # 1
        %Plug.Conn{adapter: {Plug.Cowboy.Conn, :...}, assigns: %{}, body_params: %{"session" => %{"username" => "example"}}, cookies: %{}, halted: false, host: "127.0.0.1", method: "POST", owner: #PID<0.1186.0>, params: %{"session" => %{"username" => "example"}}, path_info: ["api", "sessions"], path_params: %{}, port: 7102, private: %{MalanWeb.Router => {[], %{Plug.Swoosh.MailboxPreview => ["mailbox"]}}, :before_send => [#Function<0.23023616/1 in Plug.Telemetry.call/2>, #Function<1.46227650/1 in Phoenix.LiveReloader.before_send_inject_reloader/3>], :phoenix_action => :create, :phoenix_controller => MalanWeb.SessionController, :phoenix_endpoint => MalanWeb.Endpoint, :phoenix_format => "json", :phoenix_layout => {MalanWeb.LayoutView, :app}, :phoenix_request_logger => {"request_logger", "request_logger"}, :phoenix_router => MalanWeb.Router, :phoenix_view => MalanWeb.SessionView, :plug_session_fetch => #Function<1.77458138/1 in Plug.Session.fetch_session/1>}, query_params: %{}, query_string: "", remote_ip: {172, 27, 0, 1}, req_cookies: %{}, req_headers: [{"accept", "application/json"}, {"content-length", "34"}, {"content-type", "application/json"}, {"host", "127.0.0.1:7102"}, {"user-agent", "curl/7.64.1"}], request_path: "/api/sessions", resp_body: nil, resp_cookies: %{}, resp_headers: [{"cache-control", "max-age=0, private, must-revalidate"}, {"x-request-id", "FuK_kuIHJzqhNHcAAAOC"}], scheme: :http, script_name: [], secret_key_base: :..., state: :unset, status: nil}
    
        # 2
        %{"session" => %{"username" => "example"}}
    
    Attempted function clauses (showing 1 out of 1):
    
        def create(conn, -%{"session" => %{"username" => username, "password" => password} = session_opts}-)
    
        (malan 0.1.0) lib/malan_web/controllers/session_controller.ex:81: MalanWeb.SessionController.create/2
        (malan 0.1.0) lib/malan_web/controllers/session_controller.ex:1: MalanWeb.SessionController.action/2
        (malan 0.1.0) lib/malan_web/controllers/session_controller.ex:1: MalanWeb.SessionController.phoenix_controller_pipeline/2
        (phoenix 1.6.4) lib/phoenix/router.ex:355: Phoenix.Router.__call__/2
        (malan 0.1.0) lib/malan_web/endpoint.ex:1: MalanWeb.Endpoint.plug_builder_call/2
        (malan 0.1.0) lib/malan_web/endpoint.ex:1: MalanWeb.Endpoint."call (overridable 3)"/2
        (malan 0.1.0) lib/plug/debugger.ex:136: MalanWeb.Endpoint."call (overridable 4)"/2
        (malan 0.1.0) lib/malan_web/endpoint.ex:1: MalanWeb.Endpoint.call/2
        (phoenix 1.6.4) lib/phoenix/endpoint/cowboy2_handler.ex:54: Phoenix.Endpoint.Cowboy2Handler.init/4
        (cowboy 2.9.0) /app/deps/cowboy/src/cowboy_handler.erl:37: :cowboy_handler.execute/2
        (cowboy 2.9.0) /app/deps/cowboy/src/cowboy_stream_h.erl:306: :cowboy_stream_h.execute/3
        (cowboy 2.9.0) /app/deps/cowboy/src/cowboy_stream_h.erl:295: :cowboy_stream_h.request_process/3
        (stdlib 3.17.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
    

## Connection details

### Params

    %{"session" => %{"username" => "example"}}

### Request info

  * URI: http://127.0.0.1:7102/api/sessions
  * Query string: 

### Headers
  
  * accept: application/json
  * content-length: 34
  * content-type: application/json
  * host: 127.0.0.1:7102
  * user-agent: curl/7.64.1

### Session

    %{}
```

**Comment by FreedomBen** [comment]
Confirmed in dev.

Staging and Prod seem to behave as expected though:

Request:

```bash
#MALAN_ENDPOINT="http://localhost:4000"
MALAN_ENDPOINT="https://malan-staging.ameelio.org"

USERNAME='hp12345-d050d9dc-14e5-45dc-9a27-056ab5ef6c23'

curl  \
  --request POST \
  --header "Accept: application/json" \
  --header "Content-Type: application/json" \
  --data "{\"session\":{\"username\":\"${USERNAME}\",\"never_expires\":true}}" \
  "${MALAN_ENDPOINT}/api/sessions/"
```

Response (staging):

```
< HTTP/2 400                                                                                                                                                                                                                                  
< date: Mon, 04 Apr 2022 17:14:24 GMT                                                                                                                                                                                                         
< content-type: application/json; charset=utf-8                                                                                                                                                                                               
< content-length: 35                                                                                                                                                                                                                          
...
{"errors":{"detail":"Bad Request"}}
```

It looks like the reason for the difference in dev/staging/prod is the `debug_errors: true` in `config/dev.exs` .  The reason for this is that the missing parameter actually triggers an exception (due to pattern matching not finding a function with a compatible signature in the controller since the missing parameter blocks a match), and with debug_errors enabled the normal exception handler (which returns a Bad Request json) is replaced by one that provides the longer (and usually more useful) error message.

I can easily set that to false and that will make it return (just verified), although that will have the side effect of making logs less helpful should they be consulted.

another possible path we could take, would be to run the `release` build (same config as goes to prod)` for dependent projects.  I tend to lean against this solution as in order to make the database migrations stuff seamless like it is currently, it would require adding some sort of detection for "in release mode but not really" and have handling be different for that.

So to summarize IMHO the best solution I think is to just disable debug errors like this: https://github.com/FreedomBen/malan/commit/42480fe85c07fb7bf08183ddcb6cf7932570fa3c

Anyone developing on malan can easily re-enable it locally if desired.

@cainlevy what are your thoughts?  what would be your preference?

**Comment by cainlevy** [comment]
I tried switching from `malan-arm-dev` to `malan-arm`. Configuration was pretty easy via `DATABASE_URL` and `SECRET_KEY_BASE`. I'd be happy using this. Is there anything you can foresee that may cause trouble later?

**Comment by FreedomBen** [comment]
cool nope that's pretty much it!  just those two env vars is all you need.  If you hadn't already run the dev version, you would need to run the migrations manually since the prod release just starts the server.  The dev version uses this script ( https://github.com/FreedomBen/malan/blob/main/scripts/start-in-docker.sh ) which waits for the postgres container to come alive, then creates the database and runs the migrations if needed, and finally starts the server by running `mix phx.server`.  The release version just directly runs `mix phx.server` without running any migration.

Running migrations should be as easy as this though:

```bash
docker-compose run --rm -it malan mix ecto.setup
```

most of the time `mix ecto.migrate` would be all you need to do unless it's the first run.  `mix ecto.setup` is an alias for running these, which no-op if not needed:

```bash
mix ecto.create             # Create the database if needed
mix ecto.migrate            # Run all the migrations
mix run priv/repo/seeds.exs # Create root user (first admin user)
```

**Comment by cainlevy** [comment]
I'm happy considering the dev release to be for working on Malan itself
